en/lessons/basics/containers.md

```diff
@@ -0,0 +1,799 @@
+fromList
+  [('a',1),('a',2),('a',3),('a',4),('a',5),('a',6),('a',7),('a',8)
+  ,('b',1),('b',2),('b',3),('b',4),('b',5),('b',6),('b',7),('b',8)
+  ,('c',1),('c',2),('c',3),('c',4),('c',5),('c',6),('c',7),('c',8)
+  ,('d',1),('d',2),('d',3),('d',4),('d',5),('d',6),('d',7),('d',8)
+  ,('e',1),('e',2),('e',3),('e',4),('e',5),('e',6),('e',7),('e',8)
+  ,('f',1),('f',2),('f',3),('f',4),('f',5),('f',6),('f',7),('f',8)
+  ,('g',1),('g',2),('g',3),('g',4),('g',5),('g',6),('g',7),('g',8)
+  ,('h',1),('h',2),('h',3),('h',4),('h',5),('h',6),('h',7),('h',8)
+  ]
+```
+
+You're can thank me later when you ace the n-queens question on your next
+technical interview.
+
+## Maps
+
+In Haskell, maps are the "go-to" key-value store, sometimes referred to as a
+dictionary.
+
+A `Map` requires that its key has an ordering (`Ord k`), in the same way that
+`Set` required an ordering on the value. This is because the internal
+implementation of the `Map` type in Haskell is a size balanced binary tree. This
+is actually the same data structure as is used for `Set`.
+
+The `Map` data type and the functions for interacting with it are exported from
+`Data.Map` which is a module in the `containers` package. Don't worry about
+dependencies though, containers is a core library and ships with ghci.
+
+Like sets, a map can be constructed by inserting key value pairs into an empty map
+
+```haskell
+ghci> :t Map.empty
+Map.empty :: Map.Map k a
+ghci> Map.empty
+fromList []
+ghci> Map.insert "a" 'a' (Map.insert "b" 'b' (Map.insert "c" 'c' Map.empty))
+fromList [("a",'a'),("b",'b'),("c",'c')]
+```
+
+Or by creating it from a list
+
+```haskell
+ghci> Map.fromList [(4, '4'), (3, '3'), (2, '2'), (1,'1')]
+fromList [(1,'1'),(2,'2'),(3,'3'),(4,'4')]
+```
+
+## Updating Values to a Map
+
+A useful function to be aware of is `adjust`. This lets us update a value at a
+specified key, only if it exists, if not the old map is returned.
+
+```haskell
+ghci> Map.adjust (+2) "first" oneItem
+fromList [("first",3)]
+ghci> :t Map.adjust
+Map.adjust :: Ord k => (a -> a) -> k -> Map.Map k a -> Map.Map k a
+ghci> Map.adjust (+2) "second" oneItem
+fromList [("first",1)]
+```
+
+The observant reader will notice that `adjust` doesn't actually update the map.
+The second invocation of adjust returns the original `oneItem` map, this makes
+sense when you consider that all data in Haskell is immutable!
+
+## When to use Maps
+
+Maps are really great for in memory persistence of state that will need to be
+retrieved by a key of some arbitrary type. This is is because maps have great
+lookup asymptotics (`𝛰(log n)`) due to the ordering on the key values. The
+example that immediately springs to mind is session storage on the server in a
+web application. The session state can be indexed by an id that is stored in
+the cookie, and the session state can be retrieved from an in memory `Map` on
+every request. This solution doesn't scale infinitely, but you would be
+surprised how well it works!
+
```

Comment:
I’m not sure HashMaps are even worth mentioning in this supposedly introductory text. [There’s a recent post](https://oleg.fi/gists/posts/2021-05-19-dont-default-to-hashmap.html) by @phadej arguing that HashMaps are overrated...

Reply:
hmmmm, I would be interested in @Kleidukos thoughts here. Its a really small section and doesn't add that much, so I could see how it would be more confusing than beneficial.

Reply:
HashMap is still a pretty big part of this ecosystem, despite its warts. Let us keep this section for now.